client/vaultclient/vaultclient.go

```diff
@@ -0,0 +1,933 @@
+	c.client.SetToken(token)
+
+	// Restore the token in client to VaultClient's token
+	defer c.client.SetToken(c.config.Token)
+
+	// Read the consul ACL token and return the secret directly
+	return c.client.Logical().Read(path)
+}
+
+// RenewToken renews the supplied token and adds it to the min-heap so that it
+// is renewed periodically by the renewal loop. Any error returned during
+// renewal will be written to a buffered channel and the channel is returned
+// instead of an actual error. This helps the caller be notified of a renewal
+// failure asynchronously for appropriate actions to be taken.
+func (c *vaultClient) RenewToken(token string, increment int) <-chan error {
+	// Create a buffered error channel
+	errCh := make(chan error, 1)
+
+	if token == "" {
+		errCh <- fmt.Errorf("missing token")
+		return errCh
+	}
+	if increment < 1 {
+		errCh <- fmt.Errorf("increment cannot be less than 1")
+		return errCh
+	}
+
+	// Create a renewal request and indicate that the identifier in the
+	// request is a token and not a lease
+	renewalReq := &vaultClientRenewalRequest{
+		errCh:     errCh,
+		id:        token,
+		isToken:   true,
+		increment: increment,
+	}
+
+	// Perform the renewal of the token and send any error to the dedicated
+	// error channel.
+	if err := c.renew(renewalReq); err != nil {
+		errCh <- err
+	}
+
+	return errCh
+}
+
+// RenewLease renews the supplied lease identifier for a supplied duration and
+// adds it to the min-heap so that it gets renewed periodically by the renewal
+// loop. Any error returned during renewal will be written to a buffered
+// channel and the channel is returned instead of an actual error. This helps
+// the caller be notified of a renewal failure asynchronously for appropriate
+// actions to be taken.
+func (c *vaultClient) RenewLease(leaseId string, increment int) <-chan error {
+	c.logger.Printf("[DEBUG] client/vaultclient: renewing lease %q", leaseId)
+	// Create a buffered error channel
+	errCh := make(chan error, 1)
+
+	if leaseId == "" {
+		errCh <- fmt.Errorf("missing lease ID")
+		return errCh
+	}
+
+	if increment < 1 {
+		errCh <- fmt.Errorf("increment cannot be less than 1")
+		return errCh
+	}
+
+	// Create a renewal request using the supplied lease and duration
+	renewalReq := &vaultClientRenewalRequest{
+		errCh:     make(chan error, 1),
+		id:        leaseId,
+		increment: increment,
+	}
+
+	// Renew the secret and send any error to the dedicated error channel
+	if err := c.renew(renewalReq); err != nil {
+		errCh <- err
+	}
+
+	return errCh
+}
+
+// renew is a common method to handle renewal of both tokens and secret leases.
+// It invokes a token renewal or a secret's lease renewal. If renewal is
+// successful, min-heap is updated based on the duration after which it needs
+// renewal again. The next renewal time is randomly selected to avoid spikes in
+// the number of APIs periodically.
+func (c *vaultClient) renew(req *vaultClientRenewalRequest) error {
```

Comment:
What if we made renew just handle the renewing and it returns the next renew time and an error if the token can no longer be renewed since the TTL was hit. At that point we can remove the token from the heap and send the error to the channel.

The interface that would be nice is if this library could mask the client from all of the errors until it is fatal.